### ISSUE TITLE
Bug: Template cannot be Deployed when one host is under maintenace

### ISSUE THREAD

**Issue by GilGald** [vmware/pyvmomi#446]
Have a Template,2 hosts,one Data Center.
Data Center connected to the 2 hosts (you can see it in the vsphere tab)
When you are putting one host under maintenance and Trying to deploy a host you'll get an error that the host is not available.
but it should select the other working host,and it doesn't.

In the code I'm giving explicitly the working host and I still get the error.
>     clone_spec.location.host = available_host

It should select the available host,and I didn't find solution in the documentation.

**Comment by rreubenur** [comment]
Would you please share your clone method code snippet. Specially the ```RelocateSpec``` ?

This should work:

```
        relospec = vim.vm.RelocateSpec()
        relospec.datastore = datastore
        relospec.host = host
        relospec.pool = resource_pool
clone_spec = vim.vm.CloneSpec(powerOn=True, template=False, location=relospec, config=vm_config)
```

**Comment by GilGald** [comment]
Sure,take a look at function clone_vm in line 367 and line 416
https://github.com/QualiSystems/vCenterShell/blob/develop/package/cloudshell/cp/vcenter/common/vcenter/vmomi_service.py
Thanks.

**Comment by rreubenur** [comment]
I guess the _get_resource_pool() method is not returning the host but it is returning the resource_pool. And hence you are expecting an auto placement of the clone to the right host ? Am I right ?

Why don't you add a check to select the connected host ? 
```if summary.runtime.powerState='poweredOn' and not host.summary.runtime.inMaintenanceMode:```

**Comment by GilGald** [comment]
You are right that the host is empty and I get the resource pool,I added a code that checks if the host is under maintenance and selects a different host (that should work) and I got the same error.

Take a look at this fix that I tried on another branch :
https://github.com/QualiSystems/vCenterShell/blob/8f251fb0e2a24c6f04c2a9ba78f3cf54a411f07e/package/cloudshell/cp/vcenter/common/vcenter/vmomi_service.py line 455 try_run_clone


`    def try_run_clone(self, clone_params, clone_spec, dest_folder, logger, template, datacenter):
        is_host_error = False
        host_exception = None
        try:
            vm = self.clone_a_vm(clone_params, clone_spec, dest_folder, logger, template)
        except InvalidHostStateException as e:
            is_host_error = True
            host_exception = e

        if is_host_error:
            logger.info("InvalidHostStateException retrying to run again. " + e.message)
            available_host = self._get_working_host(datacenter)
            if available_host:
                clone_spec.location.host = available_host
                vm = self.clone_a_vm(clone_params, clone_spec, dest_folder, logger, template)
            else:
                raise host_exception

        return vm


    def _get_working_host(datacenter):
        host = None
        if datacenter.network is not None and datacenter.network[0].host is not None:
            res = next(ifilter(lambda x: not x.runtime.inMaintenanceMode, datacenter.network[0].host), None)
            if res:
                host = res.config.network.vnic[0].spec.ip.ipAddress
        return host
`

ignore that here I'm returning the res.config.network.vnic[0].spec.ip.ipAddress and it should be just the host. Its a fix i didn't checked in to git.

**Comment by rreubenur** [comment]
Ok, I can reproduce it ! I have two hosts and I put one host in maintenance mode. I have even check to select the host which is active but still clone() method is throwing error that, "The operation is not allowed in the current state of the host"

This issue need a fix if we are not doing anything wrong.

**Comment by GilGald** [comment]
That's exactly what I'm getting.
Do you know who's in charge of creating a fix for this issue?
Is that you?

**Comment by rreubenur** [comment]
@tianhao64 FYI

**Comment by tianhao64** [comment]
@GilGald Can you reproduce this using vSphere UI? If so, can you please list the steps here and I will file a bug against vsphere.

**Comment by GilGald** [comment]
@tianhao64  I cannot do it using the UI.
In the UI I can choose a host.

**Comment by GilGald** [comment]
@tianhao64 @rreubenur are there any updates ? If its a real bug it's currently blocking me.

**Comment by doppleware** [comment]
Hi,
Thanks for the help so far,Yes we would definitely appreciate a response on this. We have a product in production environments which this bug is currently impacting.
Thanks!
Roni

**Comment by jimbr70** [comment]
Any update eon this getting fixed? I'm a Cisco-contracted-employee and this issue as surfaced for us.  Need some action on it please.
